core/google/cloud/iterator.py

```diff
@@ -206,25 +243,66 @@ def __iter__(self):
         """The :class:`Iterator` is an iterator."""
         return self
 
-    def _update_page(self):
-        """Replace the current page.
+    def next_page(self, require_empty=True):
+        """Move to the next page in the result set.
+
+        If the current page is not empty and ``require_empty`` is :data:`True`
+        then an exception will be raised. If the current page is not empty
+        and ``require_empty`` is :data:`False`, then this will return
+        without updating the current page (and will return an ``updated``
+        value of :data:`False`).
 
-        Does nothing if the current page is non-null and has items
-        remaining.
+        If the current page **is** empty, but there are no more results,
+        sets the current page to :attr:`NO_MORE_PAGES`.
 
-        :raises: :class:`~exceptions.StopIteration` if there is no next page.
+        If the current page is :attr:`NO_MORE_PAGES`, throws an exception.
+
+        :type require_empty: bool
+        :param require_empty: (Optional) Flag to indicate if the current page
+                              must be empty before updating.
+
+        :rtype: bool
+        :returns: Flag indicated if the page was updated.
+        :raises ValueError: If ``require_empty`` is :data:`True` but the
+                            current page is not empty.
+        :raises ValueError: If the current page is :attr:`NO_MORE_PAGES`.
         """
-        if self.page is not None and self.page.remaining > 0:
-            return
-        if self.has_next_page():
-            response = self._get_next_page_response()
-            self._page = self.PAGE_CLASS(self, response)
+        if self._page is NO_MORE_PAGES:
```

Comment:
@jonparrott I don't really want to introduce this sentinel. I just didn't like that `page is None` for an unstarted iterator and `page is None` for an exhausted iterator. Any good ideas?

In terms of checking of `next_page()` can be done, I could do without the sentinel. If I didn't have the sentinel, then the `NO_MORE_PAGES` exception would correspond to `self._page is None and not self.has_next_page()`.